File: app/src/main/java/com/stho/software/nyota/views/SkyView.java
package com.stho.software.nyota.views;

import android.content.Context;
import android.graphics.Canvas;
import android.util.AttributeSet;

import com.stho.software.nyota.universe.AbstractSolarSystemElement;
import com.stho.software.nyota.universe.Constellation;
import com.stho.software.nyota.universe.IElement;
import com.stho.software.nyota.data.Settings;
import com.stho.software.nyota.universe.Special;
import com.stho.software.nyota.universe.Star;
import com.stho.software.nyota.utilities.Topocentric;
import com.stho.software.nyota.universe.Universe;

/**
 * Created by shoedtke on 07.09.2016.
 */
public class SkyView extends AbstractSkyView {

    private Universe universe;
    private IElement reference = null;

    public SkyView(Context context, AttributeSet attrs) {
        super(context, attrs);
        universe = Universe.getInstance();
    }

    public void setReference(IElement reference) {
        this.reference = reference;
        center.azimuth = reference.getPosition().azimuth;
        center.altitude = reference.getPosition().altitude;
        updateUI();
    }

    @Override
    protected Topocentric getReference() {
        return reference.getPosition();
    }

    @Override
    protected void onLoadSettings(Settings settings) {
        super.onLoadSettings(settings);
        super.displayNames = settings.displayNames;
        super.displaySymbols = false;
        super.displayMagnitude = settings.displayMagnitude;
    }

    public void setCenter(double centerAzimuth, double centerAltitude) {
        if (centerAzimuth != center.azimuth || centerAltitude != center.altitude) {
            center.azimuth = centerAzimuth;
            center.altitude = centerAltitude;
            updateUI();
        }
    }

    @Override
    protected void onDrawElements(Canvas canvas, double zoom) {
        for (Special special : universe.specials) {
            super.drawName(canvas, zoom, special);
        }
        for (Constellation constellation: universe.constellations.values()) {
            super.drawConstellation(canvas, zoom, constellation);
            if (super.displayNames) {
                super.drawName(canvas, zoom, constellation);
            }
        }
        for (Star star : universe.vip) {
            super.drawStar(canvas, zoom, star);
            if (super.displayNames) {
                super.drawName(canvas, zoom, star);
            }
        }
        for (AbstractSolarSystemElement element : universe.solarSystem.elements) {
            if (element.isPlanet()) {
                super.drawPlanet(canvas, zoom, element);
                if (super.displayNames) {
                    super.drawName(canvas, zoom, element);
                }
            }
        }
        super.drawMoon(canvas, zoom, universe.solarSystem.moon);
        super.drawSun(canvas, zoom, universe.solarSystem.sun);
        super.drawName(canvas, zoom, universe.getZenit(), "Z");
    }
}

File: app/src/main/java/com/stho/nyota/views/SkyView.kt
package com.stho.software.nyota.views

import android.content.Context
import android.util.AttributeSet
import com.stho.nyota.sky.universe.*
import com.stho.nyota.sky.utilities.Topocentric
import com.stho.nyota.ui.sky.ISkyViewSettings
import com.stho.nyota.ui.sky.SkyFragmentViewOptions
import com.stho.nyota.views.AbstractSkyView
import com.stho.nyota.views.ReferenceType

/**
 * Created by shoedtke on 07.09.2016.
 */
class SkyView(context: Context?, attrs: AttributeSet?) : AbstractSkyView(context, attrs) {

    private var universe: Universe? = null
    private var tippedElement: IElement? = null
    private var referenceElement: IElement? = null
    private var tippedConstellation: IElement? = null

    fun notifyDataSetChanged() {
        invalidate()
    }

    fun setUniverse(universe: Universe) {
        this.universe = universe
        invalidate()
    }

    fun setTippedElement(element: IElement?) {
        this.tippedElement = element
        invalidate()
    }

    fun removeTippedElement(element: IElement?) {
        if (this.tippedElement == element) {
            this.tippedElement = null
            this.tippedConstellation = null
            invalidate()
        }
    }

    fun setTippedConstellation(element: IElement?) {
        this.tippedConstellation = element
        invalidate()
    }

    fun setReferenceElement(element: IElement?, updateCenter: Boolean = true) {
        this.referenceElement = element
        if (updateCenter) {
            setCenter(element?.position)
        }
        invalidate()
    }

    override val referencePosition: Topocentric?
        get() = referenceElement?.position

    override fun onDrawElements() {

        drawLight()

        universe?.let {
            onDrawUniverse(it)
        }
        referenceElement?.let {
            onDrawElement(it, ReferenceType.Reference)
        }
        tippedConstellation?.let {
            onDrawElement(it, ReferenceType.TippedConstellation)
        }
        tippedElement?.let {
            onDrawElement(it, ReferenceType.TippedStar)
        }
        tippedPosition?.let {
            drawSensitivityArea(it)
        }
    }

    private fun onDrawUniverse(universe: Universe) {
        if (options.displayEcliptic) {
            super.drawEcliptic(Sun.eclipticFor(universe.moment))
        }
        universe.specials.forEach {
            super.drawNameOf(it)
        }
        universe.any.forEach {
            super.drawNameOf(it)
        }
        universe.constellations.values.forEach {
            super.drawConstellation(it, referenceType = ReferenceType.Default)
        }
        universe.vip.forEach {
            super.drawStar(it, referenceType = ReferenceType.Default)
        }
        universe.galaxies.values.forEach {
            super.drawGalaxy(it, referenceType = ReferenceType.Default)
        }
        universe.solarSystem.planets.forEach {
            super.drawPlanet(it)
        }
        if (options.displayTargets) {
            universe.targets.values.forEach {
                super.drawTarget(it)
            }
        }
        if (options.displaySatellites) {
            universe.satellites.values.forEach {
                super.drawSatellite(it)
            }
        }
        if (options.displayHints) {
            universe.hints.forEach {
                super.drawHint(it)
            }
        }
        super.drawMoon(universe.solarSystem.moon)
        super.drawSun(universe.solarSystem.sun)
        super.drawZenit(universe.zenit)
        super.drawNadir(universe.nadir)
    }

    private fun onDrawElement(element: IElement, referenceType: ReferenceType) {
        when (element) {
            is Star -> super.drawStar(element, referenceType)
            is Constellation -> super.drawConstellation(element, referenceType)
            is Galaxy -> super.drawGalaxy(element, referenceType)
        }
    }
}



